### PR TITLE
Super simple DHCP server implementation.

### PR DESCRIPTION
**Pull request by hugelgupf**

### CODE REVIEW

File: cmd/dhcp4server/main.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import (
+	"log"
+	"net"
+	"os"
+
+	"github.com/u-root/dhcp4/dhcp4server"
+)
+
+func main() {
+	_, sn, _ := net.ParseCIDR("192.168.1.0/24")
```
Comment (rminnich) [inline cmd/dhcp4server/main.go]: can this be an arg with a flag?

File: cmd/dhcp4server/main.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+	l, err := net.ListenPacket("udp4", ":67")
+	if err != nil {
+		panic(err)
+	}
+	defer l.Close()
+	s := dhcp4server.New(net.IP{192, 168, 0, 1}, sn, "", "")
```
Comment (rminnich) [inline cmd/dhcp4server/main.go]: can you avoid repeating the net.IP{} by using the sn from above?

File: dhcp4server/ipalloc.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+func newIPAllocator(subnet *net.IPNet) *ipAllocator {
+	return &ipAllocator{
+		subnet:    subnet,
+		allocated: make(map[uint32]struct{}),
+	}
+}
+
+func ipToUint32(ip net.IP) uint32 {
+	return binary.LittleEndian.Uint32(ip)
```
Comment (rminnich) [inline dhcp4server/ipalloc.go]: Do you want an ip.To4() here to be safe? I did a quick test and it seemed a good idea.

Comment (rminnich) [reply dhcp4server/ipalloc.go]: binary.LittleEndian.Uint32(ip.To4())

File: dhcp4server/server.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+func (s *Server) writePacket(conn net.PacketConn, addr net.Addr, p *dhcp4.Packet) error {
+	pkt, err := p.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	uaddr, ok := addr.(*net.UDPAddr)
+	if !ok {
+		return fmt.Errorf("what the fuck")
```
Comment (rminnich) [inline dhcp4server/server.go]: language!

Comment (rminnich) [reply dhcp4server/server.go]: :-)

File: dhcp4server/server.go [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+import (
+	"fmt"
+	"log"
+	"net"
+
+	"github.com/u-root/dhcp4"
+	"github.com/u-root/dhcp4/dhcp4opts"
+)
+
+type macAddr [16]byte
```
Comment (rminnich) [inline dhcp4server/server.go]: any reason for the 16?

File: dhcp4server/server.go [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+const maxMessageSize = 1500
```
Comment (rminnich) [inline dhcp4server/server.go]: this is not defined anywhere?

File: cmd/dhcp4server/main.go (hunk 2 above)
Comment (hugelgupf) [reply cmd/dhcp4server/main.go]: its own IP does not necessarily have to be in the sn from above -- it shouldn't be hard to figure this out on its own, but i don't want to deal with netlink or whatever. this was easier for testing.

File: dhcp4server/server.go (hunk 5 above)
Comment (hugelgupf) [reply dhcp4server/server.go]: slices can't be keys in Go maps -- has to be immutable.

File: dhcp4server/server.go (hunk 6 above)
Comment (hugelgupf) [reply dhcp4server/server.go]: I'm defining it -- in responsePacket() I'm telling clients to never send me msgs larger than 1500 bytes. Convenience for now.

File: cmd/dhcp4server/main.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+	l, err := net.ListenPacket("udp4", ":67")
+	if err != nil {
+		log.Fatalf("Could not listen on udp port 67: %v", err)
+	}
+	defer l.Close()
+
+	logger := log.New(os.Stdout, "", log.LstdFlags)
+
+	s := dhcp4server.New(net.ParseIP(*self), sn, "", "")
```
Comment (brho) [inline cmd/dhcp4server/main.go]: do you need to check the return value of net.ParseIP?  or can New() handle nil?

File: dhcp4server/server.go (hunk 5 above)
Comment (rminnich) [reply dhcp4server/server.go]: yeah I just was thinking I don't now of anything we use that's more than 6 ... no big deal

File: cmd/dhcp4server/main.go (hunk 7 above)
Comment (hugelgupf) [reply cmd/dhcp4server/main.go]: Let me put it like this... this server is not meant to be production quality, haha :)

File: dhcp4server/server.go (hunk 5 above)
Comment (hugelgupf) [reply dhcp4server/server.go]: iirc the dhcp spec leaves 16 bits for the mac addr?

File: dhcp4server/ipalloc.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		subnet:    subnet,
+		allocated: make(map[uint32]struct{}),
+	}
+}
+
+func ipToUint32(ip net.IP) uint32 {
+	return binary.LittleEndian.Uint32(ip.To4())
+}
+
+func (ia *ipAllocator) usable(ip net.IP) bool {
```
Comment (brho) [inline dhcp4server/ipalloc.go]: FYI, this ends up allowing you to allocate .0 and .255 as IP addresses.  I think you don't want to do that.

File: dhcp4server/ipalloc_test.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+	if got := ipa.alloc(); !got.Equal(want) {
+		t.Fatalf("third alloc() = %v, want %v", got, want)
+	}
+
+	want = net.IP{192, 168, 1, 2}
+	if got := ipa.alloc(); !got.Equal(want) {
+		t.Fatalf("fourth alloc() = %v, want %v", got, want)
+	}
+
+	ipa.free(net.IP{192, 168, 1, 3})
```
Comment (brho) [inline dhcp4server/ipalloc_test.go]: should it throw an error or something when you try to free something that wasn't allocated?  (i think that wasn't alloc'd)

**Pull request merged**